File: SampleApp/basefiles/src/main/java/com/rp/basefiles/BaseAdapter.java
package com.rp.basefiles;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.InvocationTargetException;
import java.util.List;

/**
 * Created by rahul on 14/1/18.
 */

public abstract class BaseAdapter<E, P extends IBaseAdapterPresenter<E>>
        extends RecyclerView.Adapter<BaseHolder<P>> {

    private final String TAG = BaseAdapter.class.getSimpleName();

    private P presenter;

    public BaseAdapter(P presenter) {
        this.presenter = presenter;
        this.presenter.onAttachAdapter(this);
    }

    @SuppressWarnings("unchecked")
    @Override
    public BaseHolder<P> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View inflate = LayoutInflater.from(parent.getContext()).inflate(getLayoutRes(viewType), parent, false);
        BaseHolder<P> baseHolder = null;
        try {
            Class<?> aClass = getClassHolder(viewType);
            baseHolder = (BaseHolder<P>) aClass.getConstructor(View.class).newInstance(inflate);
            baseHolder.setPresenter(presenter);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
        }
        return baseHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull BaseHolder<P> holder, int position) {
        Log.e(TAG, "onBindViewHolder: called");
        presenter.onAttach(holder.getView());
        presenter.onBind(position);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseHolder<P> holder, int position, @NonNull List<Object> payloads) {
        Log.e(TAG, "onBindViewHolder: payload called");
        if (payloads.isEmpty())
            onBindViewHolder(holder, position);
        else {
            presenter.onAttach(holder.getView());
            Log.e(TAG, "onBindViewHolder: payload Size => " + payloads.size()
                    + " position : " + position + " payload value : " + payloads.get(0));
            presenter.onBind(position, payloads.get(0));
        }
    }

    @Override
    public void onViewRecycled(@NonNull BaseHolder<P> holder) {
        super.onViewRecycled(holder);
    }

    @Override
    public int getItemCount() {
        return presenter.getCount();
    }

    @Override
    public int getItemViewType(int position) {
        return getViewType(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public P presenter() {
        return presenter;
    }

    public abstract int getLayoutRes(int viewType);
    public abstract Class<?> getClassHolder(int viewType);
    public abstract int getViewType(int position);
}

File: SampleApp/basefiles/src/main/java/com/rp/basefiles/BaseAdapter.kt
package com.rp.basefiles

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.lang.reflect.InvocationTargetException


abstract class BaseAdapter<E, P : IBaseAdapterPresenter<E>>(private val presenter: P)
    : RecyclerView.Adapter<BaseHolder<P>>() {

    private val TAG : String = BaseAdapter::class.java.simpleName

    init {
        presenter.onAttachAdapter(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseHolder<P> {
        val inflate = LayoutInflater.from(parent.context).inflate(getLayoutRes(viewType), parent, false)
        val str = getClassHolder(viewType).name
        var baseHolder: BaseHolder<P>? = null
        try {
            val aClass = Class.forName(str)
            baseHolder = aClass.getConstructor(View::class.java).newInstance(inflate) as BaseHolder<P>?
            baseHolder?.presenter = presenter
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
        return baseHolder as BaseHolder<P>
    }

    override fun onBindViewHolder(holder: BaseHolder<P>, position: Int) {
        Log.e(TAG, "onBindViewHolder: called")
        presenter.onAttach(holder.view)
        presenter.onBind(position)
    }

    override fun onBindViewHolder(holder: BaseHolder<P>, position: Int, payloads: MutableList<Any>) {
        Log.e(TAG, "onBindViewHolder: payload called")
        if (payloads.isEmpty())
            onBindViewHolder(holder, position)
        else {
            presenter.onAttach(holder.view)
            Log.e(TAG, "onBindViewHolder: payload Size => " + payloads.size
                    + " position : " + position + " payload value : " + payloads[0])
            presenter.onBind(position, payloads[0])
        }
    }

    override fun getItemCount(): Int {
        return presenter.count
    }

    override fun getItemViewType(position: Int): Int {
        return getViewType(position)
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    fun presenter(): P = presenter

    abstract fun getLayoutRes(viewType: Int): Int
    abstract fun getClassHolder(viewType: Int): Class<*>
    abstract fun getViewType(position: Int): Int
}